File: Professional WebBrowser - Step-By-Step - Part 3/[VB]-Professional WebBrowser - Step-By-Step - Part 3/VB/taapBrowser/Forms/FrmBrowser.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports Awesomium.Core
Imports WeifenLuo.WinFormsUI.Docking

Namespace taapBrowser.Forms
	Partial Public Class FrmBrowser
		Inherits DockContent

		#Region "Threading Delegates"
		Private Delegate Sub UpdateButtonImageDelegate()
		#End Region

		#Region "Settings"
		Private tabIconSize As Integer = 25
		Private home As String = "ccs-labs.com"
		#End Region

		''' <summary>
		''' Initialise the Browser Control
		''' </summary>
		Public Sub New()
			InitializeComponent()

			Me.AddressBar.Height = 30
			Me.LoadStyles() ' Load the AddressBar Styles
			Me.ResizeTheAddressBar()

			AddHandler Browser.DocumentReady, AddressOf Browser_DocumentReady
			AddHandler Browser.ShowPageInfo, AddressOf Browser_ShowPageInfo
		End Sub





		#Region "Styles"

		''' <summary>
		''' Load the AddressBar Styles
		''' </summary>
		Private Sub LoadStyles()
			Try

				LoadBackButton()

				LoadForwardButton()

				LoadStatusButton()

				LoadFavouritesButton()

				LoadRefreshButton()

				LoadSearchProviderButton()

				LoadSearchButton()

				LoadHomeButton()
			Catch e1 As Exception

				Throw
			End Try

		End Sub

		''' <summary>
		''' Load the User's Back button image or if none selected - the Default.
		''' </summary>
		Private Sub LoadBackButton()
			If AddressBar.InvokeRequired Then
				AddressBar.Invoke(New UpdateButtonImageDelegate(AddressOf LoadBackButton), Nothing)
			Else
				Me.btnBack.Image = ScaleImage(My.Resources.Alarm_Arrow_Left_icon, tabIconSize, tabIconSize)
			End If
		End Sub

		''' <summary>
		''' Load the User's Forward button image or if none selected - the Default.
		''' </summary>
		Private Sub LoadForwardButton()
			If AddressBar.InvokeRequired Then
				AddressBar.Invoke(New UpdateButtonImageDelegate(AddressOf LoadForwardButton), Nothing)
			Else
				Me.btnForward.Image = ScaleImage(My.Resources.Arrow_Right, tabIconSize, tabIconSize)
			End If
		End Sub

		''' <summary>
		''' Load the User's Security button image or if none selected - the Default.
		''' </summary>
		Private Sub LoadStatusButton()
			If AddressBar.InvokeRequired Then
				AddressBar.Invoke(New UpdateButtonImageDelegate(AddressOf LoadStatusButton), Nothing)
			Else
				Me.btnSecure.Image = ScaleImage(My.Resources.Green_check_box_with_check_mark_289x250, tabIconSize, tabIconSize)
			End If
		End Sub

		''' <summary>
		''' Load the User's Favourites button image or if none selected - the Default.
		''' </summary>
		Private Sub LoadFavouritesButton()
			If AddressBar.InvokeRequired Then
				AddressBar.Invoke(New UpdateButtonImageDelegate(AddressOf LoadFavouritesButton), Nothing)
			Else
				Me.btnFavourites.Image = ScaleImage(My.Resources.add_to_favourites, tabIconSize, tabIconSize)
			End If
		End Sub

		''' <summary>
		''' Load the User's reload button image or if none selected - the Default.
		''' </summary>
		Private Sub LoadRefreshButton()
			If AddressBar.InvokeRequired Then
				AddressBar.Invoke(New UpdateButtonImageDelegate(AddressOf LoadRefreshButton), Nothing)
			Else
				Me.btnReload.Image = ScaleImage(My.Resources.refresh, tabIconSize, tabIconSize)
			End If
		End Sub

		''' <summary>
		''' Load the User's Search provider button image or if none selected - the Default.
		''' </summary>
		Private Sub LoadSearchProviderButton()
			If AddressBar.InvokeRequired Then
				AddressBar.Invoke(New UpdateButtonImageDelegate(AddressOf LoadSearchProviderButton), Nothing)
			Else
				Me.btnSearchProvider.Image = ScaleImage(My.Resources.social_google_box_250x2501, tabIconSize, tabIconSize)
			End If
		End Sub

		''' <summary>
		''' Load the User's Search button image or if none selected - the Default.
		''' </summary>
		Private Sub LoadSearchButton()
			If AddressBar.InvokeRequired Then
				AddressBar.Invoke(New UpdateButtonImageDelegate(AddressOf LoadSearchButton), Nothing)
			Else
				Me.btnSearch.Image = ScaleImage(My.Resources.Search_Search_icon, tabIconSize, tabIconSize)
			End If
		End Sub

		''' <summary>
		''' Load the User's Home button image or if none selected - the Default.
		''' </summary>
		Private Sub LoadHomeButton()

			If AddressBar.InvokeRequired Then
				AddressBar.Invoke(New UpdateButtonImageDelegate(AddressOf LoadHomeButton), Nothing)
			Else
				Me.btnHome.Image = ScaleImage(My.Resources.homeenergy, tabIconSize, tabIconSize)
			End If
		End Sub

		#End Region

		#Region "Utilities"

		''' <summary>
		''' Allows the application to scale images so that they are suitable for the AddressBar
		''' </summary>
		''' <param name="image">
		''' Image: The Image to scale
		''' </param>
		''' <param name="maxWidth">
		''' Int: The Width to scale this image to.
		''' </param>
		''' <param name="maxHeight">
		''' Int: The Width to Scale this image to
		''' </param>
		''' <returns></returns>
		Private Shared Function ScaleImage(ByVal image As Image, ByVal maxWidth As Integer, ByVal maxHeight As Integer) As Image
			Dim ratioX = CDbl(maxWidth) / image.Width
			Dim ratioY = CDbl(maxHeight) / image.Height
			Dim ratio = Math.Min(ratioX, ratioY)

			Dim newWidth = CInt(image.Width * ratio)
			Dim newHeight = CInt(image.Height * ratio)

			Dim newImage = New Bitmap(newWidth, newHeight)
			Graphics.FromImage(newImage).DrawImage(image, 0, 0, newWidth, newHeight)
			Return newImage
		End Function

		''' <summary>
		''' Resizes the address bar so that the URL input area takes up as much room as possible horizontally.
		''' </summary>
		Private Sub ResizeTheAddressBar()
			Dim wid As Integer = Me.GetAddressBarControlsWidth()
			If Me.AddressBar.Size.Width - wid <> Me.tbAddressBox.Size.Width Then
				Me.tbAddressBox.Size = New Size((Me.AddressBar.Size.Width - (wid + 100)), Me.tbAddressBox.Size.Height)
			End If
		End Sub

		''' <summary>
		''' Calculates the current width of the AddressBar
		''' </summary>
		''' <returns>
		''' int: The current width of the AddressBar
		''' </returns>
		Private Function GetAddressBarControlsWidth() As Integer
			Dim width As Integer = 0

			Dim tsic As ToolStripItemCollection = AddressBar.Items

			For Each item As ToolStripItem In tsic
				If item.Name <> "tbAddressBox" Then
					width += item.Width
					width += item.Margin.Horizontal
				End If
			Next item
			Return width
		End Function

		#End Region

		#Region "Browser Events"

		''' <summary>
		''' Actions to run when the Browser Size Changes
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub FrmBrowser_Resize(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Resize

			Me.AddressBar.Height = tabIconSize
			Me.ResizeTheAddressBar()
			Me.AddressBar.Refresh()
		End Sub

		''' <summary>
		''' Actions to run when the Browser window is closing
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub FrmBrowser_FormClosing(ByVal sender As Object, ByVal e As FormClosingEventArgs) Handles MyBase.FormClosing
			Try
				Browser.Dispose()
			Catch e1 As Exception

				Throw
			End Try

		End Sub

		''' <summary>
		''' What should we do when a document has loaded?
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub Browser_DocumentReady(ByVal sender As Object, ByVal e As Awesomium.Core.UrlEventArgs)
			btnBack.Enabled = Browser.CanGoBack()
			btnForward.Enabled = Browser.CanGoForward()
		End Sub

		''' <summary>
		''' Show some basic page information.
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub Browser_ShowPageInfo(ByVal sender As Object, ByVal e As Awesomium.Core.PageInfoEventArgs)
			Dim res As String = "Certificate Error: " & e.Info.CertError.ToString()
			res &= Environment.NewLine
			res &= e.Info.ContentStatus.ToString()
			res &= Environment.NewLine
			res &= e.Info.SecurityStatus.ToString()
			MessageBox.Show(res)

		End Sub

		#End Region

		#Region "AddressBar Events"

		''' <summary>
		''' Enable or disable the
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub BtnBackClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnBack.Click
			Browser.GoBack()
		End Sub

		''' <summary>
		''' Navigate forward in the browser history.
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub BtnForwardClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnForward.Click
			Browser.CanGoForward()
		End Sub

		''' <summary>
		''' Fired when the Security button is clicked.
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub BtnSecureClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSecure.Click
			Browser.RequestPageInfo()
		End Sub

		''' <summary>
		''' Redownload and display the page again
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub BtnReloadClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnReload.Click
			Browser.Reload(True)
		End Sub

		''' <summary>
		''' Search for the User's string.
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub TbSearchBoxKeyUp(ByVal sender As Object, ByVal e As KeyEventArgs) Handles tbSearchBox.KeyUp
			If e.KeyCode = Keys.Return OrElse e.KeyCode = Keys.Enter AndAlso (Not String.IsNullOrWhiteSpace(tbSearchBox.Text)) Then
				Browser.Source = New Uri("https://www.google.co.uk/#q=" & tbSearchBox.Text.Replace(" ", "+"))
			End If
		End Sub

		''' <summary>
		''' Search for the users string
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub BtnSearchClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnSearch.Click
			If Not String.IsNullOrWhiteSpace(Me.tbSearchBox.Text) Then
				Browser.Source = New Uri("https://www.google.co.uk/#q=" & tbSearchBox.Text.Replace(" ", "+"))
			End If
		End Sub

		''' <summary>
		''' Reload the home page in this browser.
		''' </summary>
		''' <param name="sender"></param>
		''' <param name="e"></param>
		Private Sub BtnHomeClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnHome.Click
		  '  Browser.GoToHome(); not using this just now
			Browser.Source = New Uri(home)
		End Sub

		#End Region

	End Class
End Namespace

File: Professional WebBrowser - Step-By-Step - Part 3/[C#]-Professional WebBrowser - Step-By-Step - Part 3/C#/taapBrowser/Forms/FrmBrowser.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using Awesomium.Core;
using WeifenLuo.WinFormsUI.Docking;

namespace taapBrowser.Forms
{
    public partial class FrmBrowser : DockContent
    {

        #region Threading Delegates
        private delegate void UpdateButtonImageDelegate();
        #endregion

        #region Settings
        private int tabIconSize = 25;
        private string home = "ccs-labs.com";
        #endregion

        /// <summary>
        /// Initialise the Browser Control
        /// </summary>
        public FrmBrowser()
        {
            InitializeComponent();

            this.AddressBar.Height = 30;
            this.LoadStyles(); // Load the AddressBar Styles
            this.ResizeTheAddressBar();

            Browser.DocumentReady += Browser_DocumentReady;
            Browser.ShowPageInfo += Browser_ShowPageInfo;
        }





        #region Styles

        /// <summary>
        /// Load the AddressBar Styles
        /// </summary>
        private void LoadStyles()
        {
            try
            {

                LoadBackButton();

                LoadForwardButton();

                LoadStatusButton();

                LoadFavouritesButton();

                LoadRefreshButton();

                LoadSearchProviderButton();

                LoadSearchButton();

                LoadHomeButton();
            }
            catch (Exception)
            {

                throw;
            }

        }

        /// <summary>
        /// Load the User's Back button image or if none selected - the Default.
        /// </summary>
        private void LoadBackButton()
        {
            if (AddressBar.InvokeRequired)
            {
                AddressBar.Invoke(new UpdateButtonImageDelegate(LoadBackButton), null);
            }
            else
            {
                this.btnBack.Image = ScaleImage(global::taapBrowser.Properties.Resources.Alarm_Arrow_Left_icon, tabIconSize, tabIconSize);
            }
        }

        /// <summary>
        /// Load the User's Forward button image or if none selected - the Default.
        /// </summary>
        private void LoadForwardButton()
        {
            if (AddressBar.InvokeRequired)
            {
                AddressBar.Invoke(new UpdateButtonImageDelegate(LoadForwardButton), null);
            }
            else
            {
                this.btnForward.Image = ScaleImage(global::taapBrowser.Properties.Resources.Arrow_Right, tabIconSize, tabIconSize);
            }
        }

        /// <summary>
        /// Load the User's Security button image or if none selected - the Default.
        /// </summary>
        private void LoadStatusButton()
        {
            if (AddressBar.InvokeRequired)
            {
                AddressBar.Invoke(new UpdateButtonImageDelegate(LoadStatusButton), null);
            }
            else
            {
                this.btnSecure.Image = ScaleImage(global::taapBrowser.Properties.Resources.Green_check_box_with_check_mark_289x250, tabIconSize, tabIconSize);
            }
        }

        /// <summary>
        /// Load the User's Favourites button image or if none selected - the Default.
        /// </summary>
        private void LoadFavouritesButton()
        {
            if (AddressBar.InvokeRequired)
            {
                AddressBar.Invoke(new UpdateButtonImageDelegate(LoadFavouritesButton), null);
            }
            else
            {
                this.btnFavourites.Image = ScaleImage(global::taapBrowser.Properties.Resources.add_to_favourites, tabIconSize, tabIconSize);
            }
        }

        /// <summary>
        /// Load the User's reload button image or if none selected - the Default.
        /// </summary>
        private void LoadRefreshButton()
        {
            if (AddressBar.InvokeRequired)
            {
                AddressBar.Invoke(new UpdateButtonImageDelegate(LoadRefreshButton), null);
            }
            else
            {
                this.btnReload.Image = ScaleImage(global::taapBrowser.Properties.Resources.refresh, tabIconSize, tabIconSize);
            }
        }

        /// <summary>
        /// Load the User's Search provider button image or if none selected - the Default.
        /// </summary>
        private void LoadSearchProviderButton()
        {
            if (AddressBar.InvokeRequired)
            {
                AddressBar.Invoke(new UpdateButtonImageDelegate(LoadSearchProviderButton), null);
            }
            else
            {
                this.btnSearchProvider.Image = ScaleImage(global::taapBrowser.Properties.Resources.social_google_box_250x2501, tabIconSize, tabIconSize);
            }
        }

        /// <summary>
        /// Load the User's Search button image or if none selected - the Default.
        /// </summary>
        private void LoadSearchButton()
        {
            if (AddressBar.InvokeRequired)
            {
                AddressBar.Invoke(new UpdateButtonImageDelegate(LoadSearchButton), null);
            }
            else
            {
                this.btnSearch.Image = ScaleImage(global::taapBrowser.Properties.Resources.Search_Search_icon, tabIconSize, tabIconSize);
            }
        }

        /// <summary>
        /// Load the User's Home button image or if none selected - the Default.
        /// </summary>
        private void LoadHomeButton()
        {

            if (AddressBar.InvokeRequired)
            {
                AddressBar.Invoke(new UpdateButtonImageDelegate(LoadHomeButton), null);
            }
            else
            {
                this.btnHome.Image = ScaleImage(global::taapBrowser.Properties.Resources.homeenergy, tabIconSize, tabIconSize);
            }
        }

        #endregion

        #region Utilities

        /// <summary>
        /// Allows the application to scale images so that they are suitable for the AddressBar
        /// </summary>
        /// <param name="image">
        /// Image: The Image to scale
        /// </param>
        /// <param name="maxWidth">
        /// Int: The Width to scale this image to.
        /// </param>
        /// <param name="maxHeight">
        /// Int: The Width to Scale this image to
        /// </param>
        /// <returns></returns>
        private static Image ScaleImage(Image image, int maxWidth, int maxHeight)
        {
            var ratioX = (double)maxWidth / image.Width;
            var ratioY = (double)maxHeight / image.Height;
            var ratio = Math.Min(ratioX, ratioY);

            var newWidth = (int)(image.Width * ratio);
            var newHeight = (int)(image.Height * ratio);

            var newImage = new Bitmap(newWidth, newHeight);
            Graphics.FromImage(newImage).DrawImage(image, 0, 0, newWidth, newHeight);
            return newImage;
        }

        /// <summary>
        /// Resizes the address bar so that the URL input area takes up as much room as possible horizontally.
        /// </summary>
        private void ResizeTheAddressBar()
        {
            int wid = this.GetAddressBarControlsWidth();
            if (this.AddressBar.Size.Width - wid != this.tbAddressBox.Size.Width)
            {
                this.tbAddressBox.Size = new Size((this.AddressBar.Size.Width - (wid + 100)), this.tbAddressBox.Size.Height);
            }
        }

        /// <summary>
        /// Calculates the current width of the AddressBar
        /// </summary>
        /// <returns>
        /// int: The current width of the AddressBar
        /// </returns>
        private int GetAddressBarControlsWidth()
        {
            int width = 0;

            ToolStripItemCollection tsic = AddressBar.Items;

            foreach (ToolStripItem item in tsic)
            {
                if (item.Name != "tbAddressBox")
                {
                    width += item.Width;
                    width += item.Margin.Horizontal;
                }
            }
            return width;
        }

        #endregion

        #region Browser Events

        /// <summary>
        /// Actions to run when the Browser Size Changes
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void FrmBrowser_Resize(object sender, EventArgs e)
        {

            this.AddressBar.Height = tabIconSize;
            this.ResizeTheAddressBar();
            this.AddressBar.Refresh();
        }

        /// <summary>
        /// Actions to run when the Browser window is closing
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void FrmBrowser_FormClosing(object sender, FormClosingEventArgs e)
        {
            try
            {
                Browser.Dispose();
            }
            catch (Exception)
            {

                throw;
            }

        }

        /// <summary>
        /// What should we do when a document has loaded?
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        void Browser_DocumentReady(object sender, Awesomium.Core.UrlEventArgs e)
        {
            btnBack.Enabled = Browser.CanGoBack();
            btnForward.Enabled = Browser.CanGoForward();
        }

        /// <summary>
        /// Show some basic page information.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        void Browser_ShowPageInfo(object sender, Awesomium.Core.PageInfoEventArgs e)
        {
            string res = "Certificate Error: " + e.Info.CertError.ToString();
            res += Environment.NewLine;
            res += e.Info.ContentStatus.ToString();
            res += Environment.NewLine;
            res += e.Info.SecurityStatus.ToString();
            MessageBox.Show(res);

        }

        #endregion

        #region AddressBar Events

        /// <summary>
        /// Enable or disable the
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void BtnBackClick(object sender, EventArgs e)
        {
            Browser.GoBack();
        }

        /// <summary>
        /// Navigate forward in the browser history.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void BtnForwardClick(object sender, EventArgs e)
        {
            Browser.CanGoForward();
        }

        /// <summary>
        /// Fired when the Security button is clicked.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void BtnSecureClick(object sender, EventArgs e)
        {
            Browser.RequestPageInfo();
        }

        /// <summary>
        /// Redownload and display the page again
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void BtnReloadClick(object sender, EventArgs e)
        {
            Browser.Reload(true);
        }

        /// <summary>
        /// Search for the User's string.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void TbSearchBoxKeyUp(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Return || e.KeyCode == Keys.Enter && !string.IsNullOrWhiteSpace(tbSearchBox.Text))
                Browser.Source = new Uri("https://www.google.co.uk/#q=" + tbSearchBox.Text.Replace(" ", "+"));
        }

        /// <summary>
        /// Search for the users string
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void BtnSearchClick(object sender, EventArgs e)
        {
            if (!string.IsNullOrWhiteSpace(this.tbSearchBox.Text))
                Browser.Source = new Uri("https://www.google.co.uk/#q=" + tbSearchBox.Text.Replace(" ", "+"));
        }

        /// <summary>
        /// Reload the home page in this browser.
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void BtnHomeClick(object sender, EventArgs e)
        {
          //  Browser.GoToHome(); not using this just now
            Browser.Source = new Uri(home);
        }

        #endregion

    }
}
